zk-sdk/src/elgamal_program/proof_data/grouped_ciphertext_validity/handles_3.rs

```diff
@@ -0,0 +1,192 @@
+//! The grouped-ciphertext with 3 decryption handles validity proof instruction.
+//!
+//! A grouped-ciphertext validity proof certifies that a grouped ElGamal ciphertext is
+//! well-defined, i.e. the ciphertext can be decrypted by private keys associated with its
+//! decryption handles. To generate the proof, a prover must provide the Pedersen opening
+//! associated with the grouped ciphertext's commitment.
+//!
+//! In accordance with the SPL Token program application, the first decryption handle associated
+//! with the proof is referred to as the "source" handle, the second decryption handle is
+//! referred to as the "destination" handle, and the third decryption handle is referred to as the
+//! "auditor" handle.
+
+#[cfg(not(target_os = "solana"))]
+use {
+    crate::{
+        elgamal_program::errors::{ProofGenerationError, ProofVerificationError},
+        encryption::{
+            elgamal::ElGamalPubkey, grouped_elgamal::GroupedElGamalCiphertext,
+            pedersen::PedersenOpening,
+        },
+        sigma_proofs::grouped_ciphertext_validity::GroupedCiphertext3HandlesValidityProof,
+    },
+    bytemuck::bytes_of,
+    merlin::Transcript,
+};
+use {
+    crate::{
+        elgamal_program::proof_data::{ProofType, ZkProofData},
+        encryption::pod::{
+            elgamal::PodElGamalPubkey, grouped_elgamal::PodGroupedElGamalCiphertext3Handles,
+        },
+        sigma_proofs::pod::PodGroupedCiphertext3HandlesValidityProof,
+    },
+    bytemuck::{Pod, Zeroable},
+};
+
+/// The instruction data that is needed for the
+/// `ProofInstruction::VerifyGroupedCiphertext3HandlesValidity` instruction.
+///
+/// It includes the cryptographic proof as well as the context data information needed to verify
+/// the proof.
+#[derive(Clone, Copy, Pod, Zeroable)]
+#[repr(C)]
+pub struct GroupedCiphertext3HandlesValidityProofData {
+    pub context: GroupedCiphertext3HandlesValidityProofContext,
+
+    pub proof: PodGroupedCiphertext3HandlesValidityProof,
+}
+
+#[derive(Clone, Copy, Pod, Zeroable)]
+#[repr(C)]
+pub struct GroupedCiphertext3HandlesValidityProofContext {
+    pub source_pubkey: PodElGamalPubkey, // 32 bytes
+
+    pub destination_pubkey: PodElGamalPubkey, // 32 bytes
+
+    pub auditor_pubkey: PodElGamalPubkey, // 32 bytes
```

Comment:
and lastly here, we should decide on the same names as the other 3-handles proof